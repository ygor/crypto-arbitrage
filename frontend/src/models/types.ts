export enum ExchangeId {
  Binance = 'binance',
  Coinbase = 'coinbase',
  Kraken = 'kraken',
  Huobi = 'huobi',
  Bitstamp = 'bitstamp'
}

// Add activity log related types
export enum ActivityType {
  Info = 'Info',
  Warning = 'Warning',
  Error = 'Error',
  Success = 'Success'
}

export interface ActivityLogEntry {
  id: string;
  timestamp: Date | string;
  type: ActivityType;
  message: string;
  details?: string;
  relatedEntityId?: string;
  relatedEntityType?: string;
}

// Add exchange status related types
export interface ExchangeStatus {
  exchangeId: string;
  exchangeName: string;
  isUp: boolean;
  lastChecked: Date | string;
  responseTimeMs: number;
  additionalInfo?: string;
}

export enum ArbitrageOpportunityStatus {
  Detected = 0,
  Executing = 1,
  Executed = 2,
  Failed = 3,
  Missed = 4
}

export interface TradingPair {
  baseCurrency: string;
  quoteCurrency: string;
}

export interface ArbitrageOpportunity {
  id?: string;
  tradingPair?: TradingPair;
  buyExchangeId?: string | ExchangeId;
  sellExchangeId?: string | ExchangeId;
  buyPrice?: number;
  sellPrice?: number;
  quantity?: number;
  timestamp?: Date | string;
  status?: ArbitrageOpportunityStatus;
  potentialProfit?: number;
  spreadPercentage?: number;
  estimatedProfit?: number;
  detectedAt?: Date | string;
  spread?: number;
  effectiveQuantity?: number;
  isQualified?: boolean;
}

export interface PriceQuote {
  exchangeId: string;
  tradingPair: TradingPair;
  timestamp: Date | string;
  bestBidPrice: number;
  bestBidQuantity: number;
  bestAskPrice: number;
  bestAskQuantity: number;
  spread: number;
  spreadPercentage: number;
}

export interface TradeResult {
  id?: string;
  opportunityId?: string;
  tradingPair?: TradingPair;
  buyExchangeId?: string | ExchangeId;
  sellExchangeId?: string | ExchangeId;
  buyPrice?: number;
  sellPrice?: number;
  quantity?: number;
  timestamp?: Date | string;
  status?: TradeStatus;
  profitAmount?: number;
  profitPercentage?: number;
  fees?: number;
  executionTimeMs?: number;
}

export enum TradeStatus {
  Pending = 0,
  Executing = 1,
  Completed = 2,
  Failed = 3
}

export enum TradeType {
  Buy = "Buy",
  Sell = "Sell"
}

export interface ArbitrageTradeResult {
  opportunity: ArbitrageOpportunity;
  timestamp: Date | string;
  isSuccess: boolean;
  buyResult?: TradeResult;
  sellResult?: TradeResult;
  profitAmount: number;
  profitPercentage: number;
  errorMessage?: string;
}

export interface ArbitrageStatistics {
  startTime?: Date | string;
  endTime?: Date | string;
  totalProfit?: number;
  totalVolume?: number;
  totalFees?: number;
  averageProfit?: number;
  highestProfit?: number;
  lowestProfit?: number;
  totalOpportunitiesDetected?: number;
  totalTradesExecuted?: number;
  successfulTrades?: number;
  failedTrades?: number;
  averageExecutionTimeMs?: number;
  profitFactor?: number;
}

export interface Balance {
  exchangeId: string;
  currency: string;
  total: number;
  available: number;
  reserved: number;
  timestamp: Date | string;
}

export interface RiskProfile {
  name?: string;
  isActive?: boolean;
  minProfitPercentage?: number;
  minProfitAmount?: number;
  minimumProfitPercentage?: number;
  maxSlippagePercentage?: number;
  riskTolerance?: number;
  maxRetryAttempts?: number;
  maxSpreadVolatility?: number;
  stopLossPercentage?: number;
  dailyLossLimitPercent?: number;
  usePriceProtection?: boolean;
  maxTradeAmount?: number;
  maxAssetExposurePercentage?: number;
  maxTotalExposurePercentage?: number;
  dynamicSizingFactor?: number;
  maxCapitalPerTradePercent?: number;
  maxCapitalPerAssetPercent?: number;
  executionAggressiveness?: number;
  maxExecutionTimeMs?: number;
  orderBookDepthFactor?: number;
  cooldownPeriodMs?: number;
  maxConcurrentTrades?: number;
  tradeCooldownMs?: number;
  useAdaptiveParameters?: boolean;
  
  // Additional properties from frontend models
  maxExposurePerTradingPair?: number;
  maxTotalExposure?: number;
  verifyOpportunitiesBeforeExecution?: boolean;
  maxCapitalUtilizationPercentage?: number;
}

export interface ExchangeConfig {
  exchangeId: string;
  name?: string;
  isEnabled: boolean;
  apiKey?: string;
  apiSecret?: string;
  maxRequestsPerSecond?: number;
}

export interface ExchangePair {
  buyExchangeId: string;
  sellExchangeId: string;
}

export enum ArbitrageConfigEvaluationStrategy {
  Default = 0,
  Conservative = 1,
  Balanced = 2,
  Aggressive = 3
}

export enum ArbitrageConfigExecutionStrategy {
  Default = 0,
  Conservative = 1,
  Balanced = 2,
  Aggressive = 3
}

export interface ArbitrageConfig {
  isEnabled?: boolean;
  enabledTradingPairs?: string[];
  enabledBaseCurrencies?: string[];
  enabledQuoteCurrencies?: string[];
  enabledExchanges?: string[];
  enabledExchangePairs?: ExchangePair[];
  scanIntervalMs?: number;
  maxConcurrentScans?: number;
  autoTradeEnabled?: boolean;
  maxDailyTrades?: number;
  maxDailyVolume?: number;
  minOrderBookDepth?: number;
  useWebsockets?: boolean;
  usePolling?: boolean;
  pollingIntervalMs?: number;
  evaluationStrategy?: ArbitrageConfigEvaluationStrategy;
  executionStrategy?: ArbitrageConfigExecutionStrategy;
  
  // Additional properties from frontend models
  maxConcurrentExecutions?: number;
  maxTradeAmount?: number;
  autoExecuteTrades?: boolean;
  paperTradingEnabled?: boolean;
  minimumProfitPercentage?: number;
  maxExecutionTimeMs?: number;
  riskProfile?: RiskProfile;
  tradingPairs?: TradingPair[];
} 