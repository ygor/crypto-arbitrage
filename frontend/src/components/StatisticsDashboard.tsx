import React from 'react';
import { 
  Paper, Typography, Box, Card, CardContent,
  Divider, Stack
} from '@mui/material';
import {
  BarChart, Bar, LineChart, Line, XAxis, YAxis, CartesianGrid, 
  Tooltip, Legend, ResponsiveContainer, PieChart, Pie, Cell
} from 'recharts';
import { ArbitrageStatistics } from '../models/types';

interface StatisticsDashboardProps {
  statistics: ArbitrageStatistics;
}

const COLORS = ['#4caf50', '#f44336', '#2196f3', '#ff9800', '#9c27b0'];

// Default statistics to use if any property is undefined
const defaultStats: ArbitrageStatistics = {
  totalProfit: 0,
  totalVolume: 0,
  totalFees: 0,
  averageProfit: 0,
  highestProfit: 0,
  lowestProfit: 0,
  totalOpportunitiesDetected: 0,
  totalTradesExecuted: 0,
  successfulTrades: 0,
  failedTrades: 0,
  averageExecutionTimeMs: 0,
  profitFactor: 0,
  startTime: new Date().toISOString(),
  endTime: new Date().toISOString()
};

// Helper function to safely format numbers with toFixed
const safeToFixed = (value: any, decimals: number = 2): string => {
  if (value === undefined || value === null || isNaN(Number(value))) {
    return (0).toFixed(decimals);
  }
  return Number(value).toFixed(decimals);
};

const StatisticsDashboard: React.FC<StatisticsDashboardProps> = ({ statistics }) => {
  // Create a safe version of statistics with defaults for missing properties
  const safeStats = {
    ...defaultStats,
    ...Object.fromEntries(
      // Filter out any undefined or null values from statistics
      Object.entries(statistics || {}).filter(([_, v]) => v !== undefined && v !== null)
    )
  };

  const successRateData = [
    { name: 'Success', value: safeStats.successfulTrades },
    { name: 'Failed', value: safeStats.failedTrades }
  ];

  // Generate data for profit over time chart
  const profitChartData = [
    { name: 'Day 1', profit: (safeStats.totalProfit || 0) * 0.2 },
    { name: 'Day 2', profit: (safeStats.totalProfit || 0) * 0.35 },
    { name: 'Day 3', profit: (safeStats.totalProfit || 0) * 0.5 },
    { name: 'Day 4', profit: (safeStats.totalProfit || 0) * 0.75 },
    { name: 'Day 5', profit: (safeStats.totalProfit || 0) }
  ];

  // Calculate some derived metrics
  const totalTrades = (safeStats.successfulTrades || 0) + (safeStats.failedTrades || 0);
  const successRate = totalTrades > 0 ? ((safeStats.successfulTrades || 0) / totalTrades) * 100 : 0;
  const avgProfitPerTrade = (safeStats.successfulTrades || 0) > 0 ? (safeStats.totalProfit || 0) / (safeStats.successfulTrades || 0) : 0;
  
  return (
    <Box sx={{ flexGrow: 1, mt: 3 }}>
      <Typography variant="h5" gutterBottom>Arbitrage Statistics Dashboard</Typography>
      <Typography variant="subtitle1" gutterBottom color="text.secondary">
        {new Date(safeStats.startTime || new Date()).toLocaleDateString()} - {new Date(safeStats.endTime || new Date()).toLocaleDateString()}
      </Typography>
      
      {/* Key Metrics */}
      <Stack 
        direction={{ xs: 'column', sm: 'row' }} 
        spacing={2} 
        sx={{ mb: 4 }}
        justifyContent="space-between"
      >
        <MetricCard 
          title="Total Profit" 
          value={`$${safeToFixed(safeStats.totalProfit)}`}
          subtitle={`Avg: $${safeToFixed(avgProfitPerTrade)} per trade`}
          color="#4caf50"
        />
        <MetricCard 
          title="Opportunities" 
          value={(safeStats.totalOpportunitiesDetected || 0).toString()}
          subtitle={`${safeStats.totalTradesExecuted || 0} executed`}
          color="#2196f3"
        />
        <MetricCard 
          title="Success Rate" 
          value={`${safeToFixed(successRate, 1)}%`}
          subtitle={`${safeStats.successfulTrades || 0}/${totalTrades} trades`}
          color={successRate > 75 ? '#4caf50' : successRate > 50 ? '#ff9800' : '#f44336'}
        />
        <MetricCard 
          title="Trading Volume" 
          value={`$${safeToFixed(safeStats.totalVolume)}`}
          subtitle={`Fees: $${safeToFixed(safeStats.totalFees)}`}
          color="#9c27b0"
        />
      </Stack>
      
      {/* Charts */}
      <Stack spacing={3}>
        {/* Charts first row */}
        <Stack 
          direction={{ xs: 'column', md: 'row' }} 
          spacing={3}
        >
          {/* Success Rate Pie Chart */}
          <Paper elevation={2} sx={{ p: 2, height: '100%', flex: 1 }}>
            <Typography variant="h6" gutterBottom>Trade Results</Typography>
            <Box sx={{ height: 300 }}>
              <ResponsiveContainer width="100%" height="100%">
                <PieChart>
                  <Pie
                    dataKey="value"
                    data={successRateData}
                    cx="50%"
                    cy="50%"
                    outerRadius={80}
                    label={({ name, percent }) => `${name}: ${safeToFixed(percent * 100, 0)}%`}
                  >
                    {successRateData.map((entry, index) => (
                      <Cell key={`cell-${index}`} fill={index === 0 ? '#4caf50' : '#f44336'} />
                    ))}
                  </Pie>
                  <Tooltip formatter={(value) => [`${value} trades`, '']} />
                  <Legend />
                </PieChart>
              </ResponsiveContainer>
            </Box>
          </Paper>
          
          {/* Profit Over Time Line Chart */}
          <Paper elevation={2} sx={{ p: 2, flex: 2 }}>
            <Typography variant="h6" gutterBottom>Profit Over Time</Typography>
            <Box sx={{ height: 300 }}>
              <ResponsiveContainer width="100%" height="100%">
                <LineChart
                  data={profitChartData}
                  margin={{ top: 5, right: 30, left: 20, bottom: 5 }}
                >
                  <CartesianGrid strokeDasharray="3 3" />
                  <XAxis dataKey="name" />
                  <YAxis label={{ value: 'Profit ($)', angle: -90, position: 'insideLeft' }} />
                  <Tooltip formatter={(value) => [`$${safeToFixed(Number(value))}`, 'Profit']} />
                  <Legend />
                  <Line type="monotone" dataKey="profit" stroke="#4caf50" activeDot={{ r: 8 }} />
                </LineChart>
              </ResponsiveContainer>
            </Box>
          </Paper>
        </Stack>
        
        {/* Profit Metrics */}
        <Paper elevation={2} sx={{ p: 2 }}>
          <Typography variant="h6" gutterBottom>Profit Metrics</Typography>
          <Box sx={{ height: 300 }}>
            <ResponsiveContainer width="100%" height="100%">
              <BarChart
                data={[
                  { name: 'Average Profit', value: safeStats.averageProfit },
                  { name: 'Highest Profit', value: safeStats.highestProfit },
                  { name: 'Lowest Profit', value: safeStats.lowestProfit }
                ]}
                margin={{ top: 5, right: 30, left: 20, bottom: 5 }}
              >
                <CartesianGrid strokeDasharray="3 3" />
                <XAxis dataKey="name" />
                <YAxis label={{ value: 'Amount ($)', angle: -90, position: 'insideLeft' }} />
                <Tooltip formatter={(value) => [`$${safeToFixed(Number(value))}`, '']} />
                <Legend />
                <Bar dataKey="value" fill="#2196f3">
                  {[
                    <Cell key="cell-0" fill="#2196f3" />,
                    <Cell key="cell-1" fill="#4caf50" />,
                    <Cell key="cell-2" fill={(safeStats.lowestProfit || 0) < 0 ? '#f44336' : '#ff9800'} />
                  ]}
                </Bar>
              </BarChart>
            </ResponsiveContainer>
          </Box>
        </Paper>
      </Stack>
      
      {/* Additional Statistics */}
      <Paper elevation={2} sx={{ p: 2, mt: 3 }}>
        <Typography variant="h6" gutterBottom>Performance Metrics</Typography>
        <Stack
          direction={{ xs: 'column', md: 'row' }}
          spacing={2}
          justifyContent="space-between"
          divider={<Divider orientation="vertical" flexItem />}
        >
          <Box sx={{ flex: 1 }}>
            <Box sx={{ py: 1 }}>
              <Typography variant="subtitle2" color="text.secondary">Average Execution Time</Typography>
              <Typography variant="h6">{safeToFixed(safeStats.averageExecutionTimeMs)} ms</Typography>
            </Box>
            <Divider />
            <Box sx={{ py: 1 }}>
              <Typography variant="subtitle2" color="text.secondary">Profit Factor</Typography>
              <Typography variant="h6">{safeToFixed(safeStats.profitFactor)}</Typography>
            </Box>
          </Box>
          
          <Box sx={{ flex: 1 }}>
            <Box sx={{ py: 1 }}>
              <Typography variant="subtitle2" color="text.secondary">Total Opportunities</Typography>
              <Typography variant="h6">{safeStats.totalOpportunitiesDetected}</Typography>
            </Box>
            <Divider />
            <Box sx={{ py: 1 }}>
              <Typography variant="subtitle2" color="text.secondary">Total Trades</Typography>
              <Typography variant="h6">{safeStats.totalTradesExecuted}</Typography>
            </Box>
          </Box>
          
          <Box sx={{ flex: 1 }}>
            <Box sx={{ py: 1 }}>
              <Typography variant="subtitle2" color="text.secondary">Successful Trades</Typography>
              <Typography variant="h6">{safeStats.successfulTrades}</Typography>
            </Box>
            <Divider />
            <Box sx={{ py: 1 }}>
              <Typography variant="subtitle2" color="text.secondary">Failed Trades</Typography>
              <Typography variant="h6">{safeStats.failedTrades}</Typography>
            </Box>
          </Box>
        </Stack>
      </Paper>
    </Box>
  );
};

interface MetricCardProps {
  title: string;
  value: string;
  subtitle?: string;
  color?: string;
}

const MetricCard: React.FC<MetricCardProps> = ({ title, value, subtitle, color = '#000' }) => {
  return (
    <Card elevation={2} sx={{ flex: 1, minWidth: { xs: '100%', sm: 0 } }}>
      <CardContent>
        <Typography variant="subtitle1" color="text.secondary" gutterBottom>
          {title}
        </Typography>
        <Typography variant="h4" sx={{ color }}>
          {value}
        </Typography>
        {subtitle && (
          <Typography variant="body2" color="text.secondary">
            {subtitle}
          </Typography>
        )}
      </CardContent>
    </Card>
  );
};

export default StatisticsDashboard; 