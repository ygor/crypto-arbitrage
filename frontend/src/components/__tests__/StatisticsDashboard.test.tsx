import React from 'react';
import { render, screen } from '@testing-library/react';
import StatisticsDashboard from '../StatisticsDashboard';
import { ArbitrageStatistics } from '../../models/types';

// Mock recharts components to avoid SVG rendering issues in tests
jest.mock('recharts', () => {
  const OriginalModule = jest.requireActual('recharts');
  
  return {
    ...OriginalModule,
    ResponsiveContainer: ({ children }: any) => <div>{children}</div>,
    LineChart: ({ children }: any) => <div data-testid="line-chart">{children}</div>,
    BarChart: ({ children }: any) => <div data-testid="bar-chart">{children}</div>,
    PieChart: ({ children }: any) => <div data-testid="pie-chart">{children}</div>,
    Line: () => <div data-testid="line" />,
    Bar: () => <div data-testid="bar" />,
    Pie: () => <div data-testid="pie" />,
    XAxis: () => <div data-testid="x-axis" />,
    YAxis: () => <div data-testid="y-axis" />,
    CartesianGrid: () => <div data-testid="cartesian-grid" />,
    Tooltip: () => <div data-testid="tooltip" />,
    Legend: () => <div data-testid="legend" />,
    Cell: () => <div data-testid="cell" />,
  };
});

// Sample complete statistics data for testing
const mockCompleteStats: ArbitrageStatistics = {
  startTime: '2023-01-01T00:00:00Z',
  endTime: '2023-01-31T23:59:59Z',
  totalProfit: 1000.50,
  totalVolume: 50000.00,
  totalFees: 75.25,
  averageProfit: 100.05,
  highestProfit: 250.00,
  lowestProfit: 25.00,
  totalOpportunitiesDetected: 150,
  totalTradesExecuted: 25,
  successfulTrades: 20,
  failedTrades: 5,
  averageExecutionTimeMs: 125.5,
  profitFactor: 3.75
};

// Sample partial statistics with some missing data
const mockPartialStats: Partial<ArbitrageStatistics> = {
  totalProfit: 500.25,
  totalVolume: 20000.00,
  successfulTrades: 10,
  failedTrades: 2
  // Many fields are missing
};

// Empty/minimal statistics
const mockEmptyStats: Partial<ArbitrageStatistics> = {
  // Only required fields with minimal values
  startTime: new Date().toISOString(),
  endTime: new Date().toISOString(),
  totalProfit: 0,
  totalVolume: 0,
  totalFees: 0,
  averageProfit: 0,
  highestProfit: 0,
  lowestProfit: 0,
  totalOpportunitiesDetected: 0,
  totalTradesExecuted: 0,
  successfulTrades: 0,
  failedTrades: 0,
  averageExecutionTimeMs: 0,
  profitFactor: 0
};

// Extreme values for testing boundary conditions
const mockExtremeStats: ArbitrageStatistics = {
  ...mockCompleteStats,
  totalProfit: 9999999.99,
  averageExecutionTimeMs: 0.001,
  profitFactor: 9999.99
};

describe('StatisticsDashboard Component', () => {
  // Test 1: Component renders with complete data
  test('renders correctly with complete statistics data', () => {
    render(<StatisticsDashboard statistics={mockCompleteStats} />);
    
    // Check for key metrics
    expect(screen.getByText(/Total Profit/i)).toBeInTheDocument();
    expect(screen.getByText('$1000.50')).toBeInTheDocument();
    expect(screen.getByText(/Success Rate/i)).toBeInTheDocument();
    expect(screen.getByText('80.0%')).toBeInTheDocument(); // 20 successful out of 25 = 80%
  });

  // Test 2: Component handles partial data gracefully
  test('renders correctly with partial statistics data', () => {
    // Cast partial stats to full type as the component expects the full type
    render(<StatisticsDashboard statistics={mockPartialStats as ArbitrageStatistics} />);
    
    // Check that it displays the available data
    expect(screen.getByText(/Total Profit/i)).toBeInTheDocument();
    expect(screen.getByText('$500.25')).toBeInTheDocument();
    
    // Check that it gracefully handles missing data with defaults
    // This test would have failed before our fix with TypeError
    expect(screen.getByText('0.00 ms')).toBeInTheDocument(); // Default for averageExecutionTimeMs
  });

  // Test 3: Component handles empty/minimal data
  test('renders correctly with empty statistics data', () => {
    render(<StatisticsDashboard statistics={mockEmptyStats as ArbitrageStatistics} />);
    
    // Verify no errors and defaults are shown
    expect(screen.getByText(/Total Profit/i)).toBeInTheDocument();
    // Use getAllByText since there are multiple elements with $0.00 (total profit, average profit, etc.)
    const zeroAmountElements = screen.getAllByText('$0.00');
    expect(zeroAmountElements.length).toBeGreaterThan(0);
    expect(screen.getByText('0.0%')).toBeInTheDocument(); // Default success rate
  });

  // Test 4: Component handles undefined values without crashing
  test('handles undefined statistics values without errors', () => {
    // Create a minimal valid stats object, then add undefined values
    const undefinedStats = {
      startTime: '2023-01-01T00:00:00Z',
      endTime: '2023-01-31T23:59:59Z',
      totalProfit: undefined,
      totalVolume: 0,
      totalFees: 0,
      averageProfit: 0,
      highestProfit: 0,
      lowestProfit: 0,
      totalOpportunitiesDetected: 0,
      totalTradesExecuted: 0,
      successfulTrades: 0,
      failedTrades: 0,
      averageExecutionTimeMs: undefined,
      profitFactor: 0
    };
    
    // This would have crashed before our fix
    render(<StatisticsDashboard statistics={undefinedStats as any} />);
    
    // Check that it renders without errors
    expect(screen.getByText(/Total Profit/i)).toBeInTheDocument();
    // Check for default values for undefined props
    const zeroAmountElements = screen.getAllByText('$0.00');
    expect(zeroAmountElements.length).toBeGreaterThan(0);
  });

  // Test 5: Component handles extreme values
  test('renders correctly with extreme statistics values', () => {
    render(<StatisticsDashboard statistics={mockExtremeStats} />);
    
    // Check that extreme values are displayed correctly
    expect(screen.getByText('$9999999.99')).toBeInTheDocument();
    expect(screen.getByText('0.00 ms')).toBeInTheDocument();
  });

  // Test 6: Creates snapshot for regression testing
  test('matches snapshot', () => {
    const { asFragment } = render(<StatisticsDashboard statistics={mockCompleteStats} />);
    expect(asFragment()).toMatchSnapshot();
  });
}); 