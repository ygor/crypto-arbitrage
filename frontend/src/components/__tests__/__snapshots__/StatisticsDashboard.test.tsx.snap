// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticsDashboard Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-12d4nr8"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
    >
      Arbitrage Statistics Dashboard
    </h5>
    <h6
      class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom css-c268p-MuiTypography-root"
    >
      1/1/2023 - 2/1/2023
    </h6>
    <div
      class="MuiStack-root css-m6b06v-MuiStack-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 MuiCard-root css-fsggjc-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom css-c268p-MuiTypography-root"
          >
            Total Profit
          </h6>
          <h4
            class="MuiTypography-root MuiTypography-h4 css-5ksgce-MuiTypography-root"
          >
            $1000.50
          </h4>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            Avg: $50.02 per trade
          </p>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 MuiCard-root css-fsggjc-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom css-c268p-MuiTypography-root"
          >
            Opportunities
          </h6>
          <h4
            class="MuiTypography-root MuiTypography-h4 css-nyjai7-MuiTypography-root"
          >
            150
          </h4>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            25 executed
          </p>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 MuiCard-root css-fsggjc-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom css-c268p-MuiTypography-root"
          >
            Success Rate
          </h6>
          <h4
            class="MuiTypography-root MuiTypography-h4 css-5ksgce-MuiTypography-root"
          >
            80.0%
          </h4>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            20/25 trades
          </p>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 MuiCard-root css-fsggjc-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom css-c268p-MuiTypography-root"
          >
            Trading Volume
          </h6>
          <h4
            class="MuiTypography-root MuiTypography-h4 css-i0231a-MuiTypography-root"
          >
            $50000.00
          </h4>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            Fees: $75.25
          </p>
        </div>
      </div>
    </div>
    <div
      class="MuiStack-root css-v2woy3-MuiStack-root"
    >
      <div
        class="MuiStack-root css-1n7gex2-MuiStack-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-1oxpyaw-MuiPaper-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
          >
            Trade Results
          </h6>
          <div
            class="MuiBox-root css-vycneo"
          >
            <div>
              <div
                data-testid="pie-chart"
              >
                <div
                  data-testid="pie"
                />
                <div
                  data-testid="tooltip"
                />
                <div
                  data-testid="legend"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-8xn5kl-MuiPaper-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
          >
            Profit Over Time
          </h6>
          <div
            class="MuiBox-root css-vycneo"
          >
            <div>
              <div
                data-testid="line-chart"
              >
                <div
                  data-testid="cartesian-grid"
                />
                <div
                  data-testid="x-axis"
                />
                <div
                  data-testid="y-axis"
                />
                <div
                  data-testid="tooltip"
                />
                <div
                  data-testid="legend"
                />
                <div
                  data-testid="line"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-ad7o3g-MuiPaper-root"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
        >
          Profit Metrics
        </h6>
        <div
          class="MuiBox-root css-vycneo"
        >
          <div>
            <div
              data-testid="bar-chart"
            >
              <div
                data-testid="cartesian-grid"
              />
              <div
                data-testid="x-axis"
              />
              <div
                data-testid="y-axis"
              />
              <div
                data-testid="tooltip"
              />
              <div
                data-testid="legend"
              />
              <div
                data-testid="bar"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-a89azy-MuiPaper-root"
    >
      <h6
        class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
      >
        Performance Metrics
      </h6>
      <div
        class="MuiStack-root css-ccgw5h-MuiStack-root"
      >
        <div
          class="MuiBox-root css-1rr4qq7"
        >
          <div
            class="MuiBox-root css-1jpm9pd"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-hav9kw-MuiTypography-root"
            >
              Average Execution Time
            </h6>
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              125.50 ms
            </h6>
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
          />
          <div
            class="MuiBox-root css-1jpm9pd"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-hav9kw-MuiTypography-root"
            >
              Profit Factor
            </h6>
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              3.75
            </h6>
          </div>
        </div>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
        />
        <div
          class="MuiBox-root css-1rr4qq7"
        >
          <div
            class="MuiBox-root css-1jpm9pd"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-hav9kw-MuiTypography-root"
            >
              Total Opportunities
            </h6>
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              150
            </h6>
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
          />
          <div
            class="MuiBox-root css-1jpm9pd"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-hav9kw-MuiTypography-root"
            >
              Total Trades
            </h6>
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              25
            </h6>
          </div>
        </div>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
        />
        <div
          class="MuiBox-root css-1rr4qq7"
        >
          <div
            class="MuiBox-root css-1jpm9pd"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-hav9kw-MuiTypography-root"
            >
              Successful Trades
            </h6>
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              20
            </h6>
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
          />
          <div
            class="MuiBox-root css-1jpm9pd"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-hav9kw-MuiTypography-root"
            >
              Failed Trades
            </h6>
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              5
            </h6>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
